src/libfsm/print/go.c

```diff
@@ -0,0 +1,273 @@
+static int
+fsm_print_gofrag(FILE *f, const struct ir *ir, const struct fsm_options *opt,
+	const char *cp,
+	int (*leaf)(FILE *, const void *state_opaque, const void *leaf_opaque),
+	const void *leaf_opaque)
+{
+	static const struct dfavm_assembler_ir zero;
+	struct dfavm_assembler_ir a;
+	struct dfavm_op_ir *op;
+
+	static const struct fsm_vm_compile_opts vm_opts = { FSM_VM_COMPILE_DEFAULT_FLAGS, FSM_VM_COMPILE_VM_V1, NULL };
+
+	assert(f != NULL);
+	assert(ir != NULL);
+	assert(opt != NULL);
+	assert(cp != NULL);
+
+	a = zero;
+
+	/* TODO: we don't currently have .opaque information attached to struct dfavm_op_ir.
+	 * We'll need that in order to be able to use the leaf callback here. */
+	(void) leaf;
+	(void) leaf_opaque;
+
+	/* TODO: we'll need to heed cp for e.g. lx's codegen */
+	(void) cp;
+
+	if (!dfavm_compile_ir(&a, ir, vm_opts)) {
+		return -1;
+	}
+
+	/*
+	 * We only output labels for ops which are branched to. This gives
+	 * gaps in the sequence for ops which don't need a label.
+	 * So here we renumber just the ones we use.
+	 */
+	{
+		uint32_t l;
+
+		l = 0;
+
+		for (op = a.linked; op != NULL; op = op->next) {
+			if (op->num_incoming > 0) {
+				op->index = l++;
+			}
+		}
+	}
+
+	for (op = a.linked; op != NULL; op = op->next) {
+		if (op->num_incoming > 0) {
+			print_label(f, op, opt);
+			fprintf(f, "\n");
+		}
+
+		fprintf(f, "\t");
+
+		switch (op->instr) {
+		case VM_OP_STOP:
+			print_cond(f, op, opt);
+			print_end(f, op, opt, op->u.stop.end_bits, ir);
+			break;
+
+		case VM_OP_FETCH:
+			print_fetch(f, opt);
+			print_end(f, op, opt, op->u.fetch.end_bits, ir);
+			break;
+
+		case VM_OP_BRANCH:
+			print_cond(f, op, opt);
+			print_branch(f, op);
+			break;
+
+		default:
+			assert(!"unreached");
+			break;
+		}
+
+		fprintf(f, "\n");
+	}
+
+	dfavm_opasm_finalize_op(&a);
+
+	return 0;
+}
+
+static void
+fsm_print_go_complete(FILE *f, const struct ir *ir, const struct fsm_options *opt)
+{
+	const char *cp;
+
+	assert(f != NULL);
+	assert(ir != NULL);
+	assert(opt != NULL);
+
+	(void) fsm_print_gofrag(f, ir, opt, cp,
+		opt->leaf != NULL ? opt->leaf : leaf, opt->leaf_opaque);
+}
+
+void
+fsm_print_go(FILE *f, const struct fsm *fsm)
+{
+	struct ir *ir;
+	const char *prefix;
+
+	assert(f != NULL);
+	assert(fsm != NULL);
+	assert(fsm->opt != NULL);
+
+	ir = make_ir(fsm);
+	if (ir == NULL) {
+		return;
+	}
+
+	/* henceforth, no function should be passed struct fsm *, only the ir and options */
+
+	if (fsm->opt->prefix != NULL) {
+		prefix = fsm->opt->prefix;
+	} else {
+		prefix = "fsm_";
+	}
+
+	if (fsm->opt->fragment) {
+		fsm_print_go_complete(f, ir, fsm->opt);
+		return;
+	}
+
+	fprintf(f, "package %sfsm\n", prefix);
+	fprintf(f, "\n");
+
+	fprintf(f, "func Match");
+
+	switch (fsm->opt->io) {
+	case FSM_IO_STR:
+		fprintf(f, "(data []byte) int {\n");
+		/* start idx at -1 unsigned so after first increment we're correct at index 0 */
+		fprintf(f, "\tvar idx = ^uint(0)\n");
+		fprintf(f, "\n");
+		break;
```

Comment:
I might move this under something other than `FSM_IO_STR` (since I'd expect that to emit go's `string` type). Generating `[]byte` is great, but I want to call it something else, and I want to come back and revisit how `FSM_IO_*` are named.